src/ipc/mod.rs

```diff
@@ -109,20 +120,20 @@ impl<'a, T: Ipc> Backend<'a, T> {
     /// Get the next IPC message.
     // This is similar to `impl Iterator`, but the returned value is tied to the lifetime
     // of `self`, so we cannot implement that trait.
-    pub fn next(&mut self) -> Option<Msg<'_>> {
+    pub fn next(&mut self) -> Option<(Msg<'_>, T::Addr)> {
         // if we have leftover buffer from the last read, parse another message.
         if self.read_until < self.tot_read {
             let (msg, consumed) = Msg::from_buf(&self.receive_buf[self.read_until..]).ok()?;
             self.read_until += consumed;
-            Some(msg)
+            Some((msg, self.last_recv_addr.clone()))
         } else {
             self.tot_read = self.get_next_read().ok()?;
             self.read_until = 0;
             let (msg, consumed) =
                 Msg::from_buf(&self.receive_buf[self.read_until..self.tot_read]).ok()?;
             self.read_until += consumed;
 
-            Some(msg)
+            Some((msg, self.last_recv_addr.clone()))
```

Comment:
Is it possible for reads from different datapath segments to interleave? Would be good to write a comment explaining why this `last_recv_addr` stuff is safe to do.

Reply:
Good call, I agree this should be documented for future reference. It should be fine because we don't actually call `get_next_read()` again until we've exhausted all the messages from the last read. But let me know if you think I'm missing something